Feature: Feedback form attachments

nbradbury opened the pull request:
This is the feature branch for uploading feedback form attachments. It contains the following child PRs:

* #21107
* #21115
* #21116 
* #21117 

To test:

1. Tap Help on the Me page and ensure the "Contact email" is correct (this for step 6)
2. Return to the Me page and tap "Send feedback"
3. Enter a message
4. Add a couple of attachments
5. Tap submit and verify it all works
6. When you get the email about the ticket, go to https://a8c.zendesk.com/agent/tickets/$ticketNumber and verify the attachments were added

Caveats:

- Unlike iOS, there is no preview of attachments yet - that's being worked on separately in [this branch](https://github.com/wordpress-mobile/WordPress-Android/pull/21121)
- Like iOS, there isn't a way to cancel the attachment uploading. I suspect we'll want to add this at some point.

nbradbury committed:
Copied some basic attachment code from DOAndroid

nbradbury committed:
Added basic attachments section and photo picker

nbradbury committed:
Rewrote preview

nbradbury committed:
Fixed attachment button

nbradbury committed:
Removed separate attachment section

nbradbury committed:
Added attachment to preview, added MediaBrowserType.FEEDBACK_FORM_SINGLE_MEDIA_PICKER

nbradbury committed:
Cleaned up FeedbackFormScreenPreview

nbradbury committed:
Fixed ReturnCount warning in addAttachment

nbradbury committed:
Fixed detekt warnings in Uri.copyToTempFile

nbradbury committed:
Fixed remaining detekt warnings in UriExt

nbradbury committed:
Added comments to UriExt

nbradbury committed:
Add single selected media attachments

nbradbury committed:
Allow multi select

nbradbury committed:
Merge branch 'feature/feedback-form-attachments' of https://github.com/wordpress-mobile/WordPress-Android into issue/21105-feedback-form-attachments-ui

nbradbury committed:
Always use showToast fun

nbradbury committed:
Fixed checkstyle issue with FeedbackFormUtils

nbradbury committed:
Fixed failing test

nbradbury committed:
Simplified attachment types to just image & video

nbradbury committed:
Include attachments in discard confirmation

nbradbury committed:
Merge pull request #21107 from wordpress-mobile/issue/21105-feedback-form-attachments-ui

Feedback form attachments UI

nbradbury committed:
Added attachment IDs

nbradbury committed:
First pass at ZendeskUploadHelper

nbradbury committed:
Fixed warnings

nbradbury committed:
Make attachmentIds optional

nbradbury committed:
Merge branch 'trunk' into feature/feedback-form-attachments

nbradbury committed:
Merge branch 'feature/feedback-form-attachments' of https://github.com/wordpress-mobile/WordPress-Android into issue/21105-feedback-form-upload-attachments

dangermattic commented:
<!--
  0 Errors
  7 Warnings: <code>strings.xml</code> files...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          7 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>strings.xml</code> files should only be updated on release branches, when the translations are downloaded by our automation.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ZendeskUploadHelper</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ProgressDialogState</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>FeedbackFormAttachment</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>FeedbackFormUtils</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

nbradbury committed:
Upload multiple attachments

nbradbury committed:
Pass attachment IDs when creating request

nbradbury committed:
Merge branch 'trunk' into feature/feedback-form-attachments

nbradbury committed:
Merge branch 'feature/feedback-form-attachments' of https://github.com/wordpress-mobile/WordPress-Android into issue/21105-feedback-form-upload-attachments

nbradbury committed:
Moved uploading attachments to the view model

nbradbury committed:
Updated to use attachment tokens

nbradbury committed:
Use tempfile rather than uri for upload

nbradbury committed:
Use a completion handler to await uploads

nbradbury committed:
Minor cleanup

nbradbury committed:
Fixed checkstyle warnings

nbradbury committed:
Merge pull request #21115 from wordpress-mobile/issue/21105-feedback-form-upload-attachments

Feedback form: upload attachments

nbradbury committed:
Improved comment

nbradbury committed:
Added simple progress dialog

nbradbury committed:
Simplified progress dialog

nbradbury committed:
Use a progress dialog during uploading

nbradbury committed:
Don't allow progress to be dismissed

nbradbury committed:
First pass at suspend upload

nbradbury committed:
Second pass at uploading in a coroutine

nbradbury committed:
Merge pull request #21116 from wordpress-mobile/issue/21105-feedback-form-progress

Feedback form progress

nbradbury committed:
Removed unused fun

nbradbury committed:
Merge branch 'feature/feedback-form-attachments' into issue/21105-feedback-form-suspend

nbradbury committed:
Fixed checkstyle error

nbradbury committed:
Reverted to suspendCoroutine

nbradbury committed:
Merge pull request #21117 from wordpress-mobile/issue/21105-feedback-form-suspend

Feedback form: suspend while uploading

nbradbury committed:
Make sure continuation is active before resuming

nbradbury committed:
Removed redundant logging

nbradbury committed:
Fixed checkstyle issue

nbradbury committed:
Merge branch 'feature/feedback-form-attachments' of https://github.com/wordpress-mobile/WordPress-Android into issue/21105-feedback-form-suspend-cancel

Conflicts:
	WordPress/src/main/java/org/wordpress/android/support/ZendeskUploadHelper.kt

nbradbury committed:
Improved preview

nbradbury committed:
Minor preview change

nbradbury committed:
Don't continue adding attachments when one fails

nbradbury committed:
Changed MAX_SINGLE_ATTACHMENT_SIZE to match iOS

nbradbury committed:
Removed max total attachment size to match iOS

nbradbury committed:
Changed MAX_ATTACHMENTS to 5 to match iOS

nbradbury committed:
Don't hard-code the size limit in the string resource

nbradbury committed:
Fixed failing test

nbradbury committed:
Merge pull request #21119 from wordpress-mobile/issue/21105-feedback-form-suspend-cancel

Feedback form: suspend cancel upload

nbradbury committed:
Merge branch 'trunk' into feature/feedback-form-attachments

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/ui/main/feedbackform/FeedbackFormScreen.kt:
I noted the message box is not auto-focused and does not enable shift key when focused (i.e., begin the input/sentence with a capital letter). My perception is that doing so is a common practice on iOS and Android. Should we consider implementing that here?

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/ui/main/feedbackform/FeedbackFormViewModel.kt:
```diff
@@ -98,13 +160,98 @@ class FeedbackFormViewModel @Inject constructor(
+            showToast(R.string.feedback_form_attachment_already_added)
+        } else if (size > MAX_ATTACHMENT_SIZE) {
+            showToast(R.string.feedback_form_attachment_too_large)
+        } else if (file == null) {
+            showToast(R.string.feedback_form_unable_to_create_tempfile)
```
I encountered this error when attempt to attach an image when "browsing" files on a Samsung Galaxy S20 running Android 13. Should this use case be supported? 

<details><summary>Screen recording</summary>

https://github.com/user-attachments/assets/b892e77a-6000-4492-a989-72d7c351e1c5

</details>

dcalhoun reviewed (approved):
Nice work!

I tested the flow in both the Jetpack and WordPress mobile apps on a Samsung Galaxy S20. I tested the happy path and disrupting submissions with network conditions. I did not encounter any significant issues. 

In the WordPress app, there is no contact email. I presume we want to allow users to submit anonymous feedback from this app. Is that accurate? 

I left a couple of suggestions, but I do not consider them blockers. Also, feel free to seek additional Android code feedback if you'd like.

nbradbury committed:
Auto-focus the message field

nbradbury committed:
Fix file:// URIs

nbradbury committed:
Corrected typo in comment

nbradbury committed:
Use KeyboardCapitalization.Sentences

nbradbury commented:
@dcalhoun Thanks for the review!

> I noted the message box is not auto-focused and does not enable shift key be focused...Should we consider implementing that here?

Agreed. I've added auto-focus but I'm not sure why sentence capitalization isn't working since I'm requesting it. ~~I'll keep looking into that, but I may file a separate issue for it to avoid holding up this PR.~~ **Update:** it turns out I was doing it right in [bd868ee](https://github.com/wordpress-mobile/WordPress-Android/pull/21114/commits/bd868eedbbf6bd7cfb14ee42baa029b04629d68a) after all, so problem solved.

> I encountered this error when attempt to attach an image when "browsing" files on a Samsung Galaxy S20

Very good catch! I was able to reproduce this when browsing on my Pixel 7a and fixed it in [ddfb08b](https://github.com/wordpress-mobile/WordPress-Android/pull/21114/commits/ddfb08b6369473f8856c54db0a046811fcc9f71d).

nbradbury commented:
> In the WordPress app, there is no contact email. I presume we want to allow users to submit anonymous feedback from this app. Is that accurate?

That's accurate, but is subject to change.
